File: 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/utils/CPSCounter.java
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.utils;

/**
 * @author superblaubeere27
 */
public class CPSCounter {
    private static final int MAX_CPS = 50;
    private static final RollingArrayLongBuffer[] TIMESTAMP_BUFFERS = new RollingArrayLongBuffer[MouseButton.values().length];

    static {
        for (int i = 0; i < TIMESTAMP_BUFFERS.length; i++) {
            TIMESTAMP_BUFFERS[i] = new RollingArrayLongBuffer(MAX_CPS);
        }
    }

    /**
     * Registers a mouse button click
     *
     * @param button The clicked button
     */
    public static void registerClick(MouseButton button) {
        TIMESTAMP_BUFFERS[button.getIndex()].add(System.currentTimeMillis());
    }

    /**
     * Gets the count of clicks that have occurrence since the last 1000ms
     *
     * @param button The mouse button
     * @return The CPS
     */
    public static int getCPS(MouseButton button) {
        return TIMESTAMP_BUFFERS[button.getIndex()].getTimestampsSince(System.currentTimeMillis() - 1000L);
    }

    public enum MouseButton {
        LEFT(0), MIDDLE(1), RIGHT(2);

        private int index;

        MouseButton(int index) {
            this.index = index;
        }

        private int getIndex() {
            return index;
        }
    }

}

File: shared/main/java/net/ccbluex/liquidbounce/utils/CPSCounter.kt
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.utils

/**
 * @author superblaubeere27
 */
object CPSCounter
{
	private const val MAX_CPS = 50
	private val TIMESTAMP_BUFFERS = arrayOfNulls<RollingArrayLongBuffer>(MouseButton.values().size)

	/**
	 * Registers a mouse button click
	 *
	 * @param button
	 * The clicked button
	 */
	@JvmStatic
	fun registerClick(button: MouseButton)
	{
		TIMESTAMP_BUFFERS[button.index]?.add(System.currentTimeMillis())
	}

	/**
	 * Gets the count of clicks that have occurrence since the last 1000ms
	 *
	 * @param  button
	 * The mouse button
	 * @return        The CPS
	 */
	fun getCPS(button: MouseButton): Int = TIMESTAMP_BUFFERS[button.index]?.getTimestampsSince(System.currentTimeMillis() - 1000L) ?: 0

	enum class MouseButton(val index: Int)
	{
		LEFT(0), MIDDLE(1), RIGHT(2);
	}

	init
	{
		for (i in TIMESTAMP_BUFFERS.indices) TIMESTAMP_BUFFERS[i] = RollingArrayLongBuffer(MAX_CPS)
	}
}
